### PR TITLE
Custom Visuals - Add OnObject formatting utils

### PR DESCRIPTION
**Pull request by shafeeqz**
Adding a utils for the onobject formatting API, 
The main class is HtmlSubSelectionHelper,
 
added some helper array functions.

I will update the Readme file

### CODE REVIEW

**Review by AleksSavelev — DISMISSED** [review]

File: package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "powerbi-visuals-utils-onobjectformatting",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.ts",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (AleksSavelev) [inline package.json]: Please, add `build` command, since we'll publish it to npmjs

File: src/types.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    displayName: string;
+    showUI: boolean;
+    selectionId?: ISelectionId;
+    selectionOrigin?: powerbi.extensibility.IPoint;
+    focusOrder?: number;
+    metadata?: unknown;
+}
+
+/**
+ * Helper tool that meakes it easier for visuals to manage sub-selection.
```
Comment (AleksSavelev) [inline src/types.ts]: Helper tool that **makes** it easier for visuals to manage sub-selection.

**Review by AleksSavelev — COMMENTED** [review]
Couple suggestions. I'll take a look one more time a little bit later

File: src/HtmlSubSelectionHelper.ts [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+
+    private getRestrictionElement(element: HTMLElement, type: SubSelectionOutlineRestrictionType): HTMLElement | undefined {
+        const restrictionElement: HTMLElement = element.closest(`[${SubSelectableRestrictingElementAttribute}="${type}"]`);
+        if (restrictionElement && helperOwnsElement(this.hostElement, restrictionElement)) {
+            return restrictionElement;
+        }
+        return undefined;
+    }
+
+    private applyElementRestriction(restrictingElement: HTMLElement, subselectionElement: HTMLElement): IRect | undefined {
```
Comment (AleksSavelev) [inline src/HtmlSubSelectionHelper.ts]: Why is "undefined" possible here? I think it should be only IRect

File: src/HtmlSubSelectionHelper.ts [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+                undefined /*event*/,
+                selectionOrigins[index],
+            )
+        ));
+        return visualSubSelections;
+    }
+
+    public createVisualSubSelectionForSingleObject(createVisualSubSelectionArgs: CreateVisualSubSelectionFromObjectArgs): CustomVisualSubSelection {
+        const { objectName, subSelectionType, displayName, showUI, selectionId, selectionOrigin, focusOrder, metadata } = createVisualSubSelectionArgs;
+        //const { top } = (target as HTMLElement)?.getBoundingClientRect() ?? {top: 0};
```
Comment (AleksSavelev) [inline src/HtmlSubSelectionHelper.ts]: I guess comment line is redundant

File: src/HtmlSubSelectionHelper.ts [hunk 5]
```diff
@@ -0,0 +1,830 @@
...
+        const origin = (selectionOrigin && subSelectionType in [SubSelectionStylesType.Text, SubSelectionStylesType.NumericText])
+            ? { ...selectionOrigin, offset: { x: 0, y: (selectionOrigin?.y - 0) * -1 } }
+            : selectionOrigin;
```
Comment (AleksSavelev) [inline src/HtmlSubSelectionHelper.ts]: These lines looks pretty complicated. Let's split it up a little bit.
``` typescript
const useOffsetInSelection = selectionOrigin && subSelectionType in [SubSelectionStylesType.Text, SubSelectionStylesType.NumericText];
const origin = useOffsetInSelection ? { ...selectionOrigin, offset: { x: 0, y: (selectionOrigin?.y - 0) * -1 } } : selectionOrigin;
```

File: src/HtmlSubSelectionHelper.ts [hunk 6]
```diff
@@ -0,0 +1,830 @@
...
+        }
+        return Number(type);
+    }
+
+    private getSubSelectableElements(): HTMLElement[] {
+        const hostElement = this.hostElement;
+        return this.host.selectAll<HTMLElement, unknown>(HtmlSubSelectableSelector).filter(function () {
+            const element = this;
+            return helperOwnsElement(hostElement, element);
+        }).nodes();
```
Comment (AleksSavelev) [inline src/HtmlSubSelectionHelper.ts]: Don't you mind to change the whitespaces here? Just to make it easier to read
``` typescript
     return this.host
            .selectAll<HTMLElement, unknown>(HtmlSubSelectableSelector)
            .filter(function () {
                const element = this;
                return helperOwnsElement(hostElement, element);
            })
            .nodes();
```

**Review by AleksSavelev — CHANGES_REQUESTED** [review]

File: package.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "powerbi-visuals-utils-onobjectformatting",
+  "version": "1.0.0",
```
Comment (AleksSavelev) [inline package.json]: All our utils is 6+, let's make it same (6.0.0). It makes easier to track the latest versions

File: package.json (hunk 1 above)
Comment (shafeeqz) [reply package.json]: added.

File: src/types.ts (hunk 2 above)
Comment (shafeeqz) [reply src/types.ts]: done.

File: src/HtmlSubSelectionHelper.ts (hunk 3 above)
Comment (shafeeqz) [reply src/HtmlSubSelectionHelper.ts]: removed.

File: src/HtmlSubSelectionHelper.ts (hunk 4 above)
Comment (shafeeqz) [reply src/HtmlSubSelectionHelper.ts]: done.

File: src/HtmlSubSelectionHelper.ts (hunk 5 above)
Comment (shafeeqz) [reply src/HtmlSubSelectionHelper.ts]: done.

File: src/HtmlSubSelectionHelper.ts (hunk 6 above)
Comment (shafeeqz) [reply src/HtmlSubSelectionHelper.ts]: done.

File: package.json (hunk 7 above)
Comment (shafeeqz) [reply package.json]: done.

**Review by AleksSavelev — APPROVED** [review]

**Pull request merged**